### DISCUSSION TITLE
Stargate Code Refactoring (v1->v2)

### DISCUSSION THREAD

**Discussion by jeffreyscarpenter** [stargate/stargate#1454] [RFCs]
# Stargate Code Refactoring (v1->v2)
This document provides an analysis of the Maven modules in Stargate v1 and proposes some refactorings for Stargate v2.

## Background (V1)

The existing Stargate repo consists of about 20 modules, consisting of:

* 3 Persistence implementations (for C\* 3.11, C\* 4.0, DSE)
* 4 API modules (cql, grpc, graphql, rest/doc) -- REST and Document APIs co-located due to historical reasons
* 2 Test/testing-related modules (testing, testing-services)
* Starter module “stargate-starter” that bootstraps the whole service via OSGi
* Shared modules
    * 4 Authnz-related shared modules
    * 2 Config-related shared modules
    * Persistence-common shared by persistence implementations
    * Persistence-api shared by… APIs and persistence
    * “Core” shared by almost everything (potentially)
    * Health-checker and metrics-jersey are used by HTTP based APIs
    * Rate-limiting-global is intended as an extension point.

Due to the monolithic nature of the V1 implementation, there are many dependencies across components.

## Problems for V2

Stargate V2 effort will decouple the current “Persistence” part (into New Coordinator entity, consisting of Bridge along with Persistence-backed Coordinator) from APIs/Services.

Services will also be decoupled from each other, either living in separate repos, or at least as properly separated share-little sub-modules.

This means that some of the code-level sharing that currently exists needs to be severed: no code shared by being in the same repo (as is the case with REST/Doc API), all sharing via explicit “common” packages.

The current placement of code that is shared by multiple modules is arbitrary, both leading to duplication where functionality is not shared (despite making sense) and sometimes shared in the wrong scope (or sharing things for which sharing adds otherwise unnecessary and awkward dependency -- currently the case for REST/Doc API pair, due to co-location).

### Java 9+: No “split” Java packages

Starting with Java 9 (or more accurately, Java module system JPMS), Java packages cannot be split across multiple jars/bundles/modules: something that is allowed up until Java 8. \
Currently known problems (Split packages) include:

* - none observed yet -

### Dependency to Jersey framework (and maybe DropWizard)

Some of Metrics-registration code as well as wiring relies on Jersey framework. There is obviously a need for wiring for the current implementation. But if we are to consider moving to newer frameworks (especially for APIs/Services), we need to try to minimize them to make moving easier. And at very least we need to be aware what the dependencies are.

## Scopes for shared functionality

Considering existing functionality, there seem to be different “scopes” for functionality sharing. \
Scopes are defined in the context of V2 view, in which “Persistence” (which refers to the “Coordinator” entity, also including Bridge) is physically separate from Services (APIs). This separation exists mostly as a logical concern in V1.

What follows is a combination of existing state, ideas for grouping shared code along these scopes.

### Functionality shared by Persistence implementations

Currently “persistence-common” seems to align with this scope.

The gRPC Bridge component will be packaged in a similar way to the gRPC backend

* Access to authnz?

### Functionality shared by HTTP-based Services

Currently not well defined; existing “restapi” contains code that is shared between Doc and REST APIs due to their incidental co-location.

Misc ideas, questions:

* ApiError type (renamed from “Error”, exposed by REST/Doc API, see ​​[https://github.com/stargate/stargate/issues/1322](https://github.com/stargate/stargate/issues/1322) )
* RxJava helpers?
* Remaining things shared by Doc/REST APIs?
* Do we need this as a separate scope, or just accept some bloat wrt non-http services’ dependencies?

### Functionality shared by all Services/APIs (but not Persistence)

Misc ideas:

* QueryBuilder
* Health-checker (should mostly disappear in v2)
* gRPC Client
* Front-end metrics (if different from Persistence)

### Functionality shared by Persistence and Services (global)

Currently we have “core” and likely “persistence-api”. Ideally there would be few things we really want to share across all main-level entities but perhaps some cross-cutting concerns should be shared:

* Metrics?

# Proposed New/Changed/Renamed/Removed Shared Modules

This section provides detailed recommendations by module based on the principles above.

## Common modules 

Shared by both service-side and persistence-side.

### `grpc-proto` module

This module contains the GRPC protos as well as the generated Java client/server stubs. There are repos for other languages: 

* [https://github.com/stargate/stargate-grpc-java-client](https://github.com/stargate/stargate-grpc-java-client) -> should we deprecate this? Commit messages indicate this was an experiment in stub generation and contains no other contents. Actual Java stubs we use are in this module (grpc-proto)
* [https://github.com/stargate/stargate-grpc-rust-client](https://github.com/stargate/stargate-grpc-rust-client) - contains generated stub (S?) and a bit of wrapper code. Why does it duplicate proto files from this module?
* [https://github.com/stargate/stargate-grpc-go-client](https://github.com/stargate/stargate-grpc-go-client)  - contains generated stub (S?) and a bit of wrapper code. Does it duplicate proto files or pull from this location?

Current recommendation is to leave these stubs for other languages in their own repos.

### `core` module

Contains 3 packages

* Io.stargate.core.activator - base classes for working with OSGI. Can potentially move to a coordinator common as we eliminate OSGI from service APIs
* Io.stargate.core.metrics - used in multiple locations. Would this be affected by any planned changes in metrics approach?
* Io.stargate.core.util - ByteBufferUtils used by rest-api (including both REST and docs API) and graphqlapi (minimally). Could move to service-common

### `authnz` module

Core classes for authentication and authorization used in multiple persistence implementations and other locations in v1. Based on design decision for v2 to have authentication performed by the Bridge rather than services, this could potentially be aligned with a persistence-common (see below). No changes required for v2.

### `health-checker` module

Includes health checks for both coordinator and service side modules. Each checker class should be refactored to be as close to where it is used as possible. Each Service will have its own health checker. Checks for OSGi state so in current form are mostly relevant for the coordinator side.

Note also that unit tests are not packaged separately in this module. Unit tests should be moved to an appropriate location in the destination module as part of any refactoring we do.

### `persistence-api` module 

Goal: eliminate the dependence of Service API modules on `persistence-api`

Recommend movement of `QueryBuilder` (basically, the entire `io.stargate.db.query` package) to `service-common`, as this is not used in persistence implementations. (Alternatively, `QueryBuilder` could be replaced entirely).

`Io.stargate.db` contains many classes forming a common language (Table, Column, etc) -> perhaps some of this should move to `persistence-common`?

`Io.stargate.db.limiter` - only used by `rate-limiting-global` (which in turn is currently unused) -> consider moving to `persistence-common`? 

### `persistence-common` module

This is indeed only used by other persistence modules. (There may be items we move to here as part of other cleanup?)

This module has a dependency on the cassandra-all module (4.0) which we should try to eliminate, we don’t want a dependency to a specific Cassandra version.

### `bridge-proto` module (new)

Defer creating until we need, probably by copy-paste from `grpc-proto` to start.

### `bridge` module (new)

Defer creating until we need, probably by copy paste from `grpc`.

### `cql` module

No changes required for v2.

### `persistence-cassandra-3.11` module

No changes required for v2.

### `persistence-cassandra-4.0` module

No changes required for v2.

### `persistence-common` module

Currently contains classes that help manage Cassandra-based persistence

TBD - could be a landing place for items we move out of persistence-api?

### `rate-limiting-global` module

Contains a single class - the only Implementation of an interface defined in persistence-api. Notes indicate this is not intended for production use. No changes required for v2.

### `auth-api` module

This service provides the authentication endpoint clients use to get credentials. This would continue to be part of the coordinator node because the table-based authentication implementation needs to interact with the selected persistence backend. No changes required for v2.

### `auth-jwt-service` module 

Provides implementations of the `AuthenticationService` and `AuthorizationService` defined in `authnz`. 

This would continue to be part of the coordinator node. No changes required for v2, but I think we should consider refactoring auth handling to support a chain instead of a single implementation. That way you can register your auth provider with the bridge and configure the order instead of picking a particular implementation.

The dependencies on the persistence-api include the `o.a.c.stargate.exceptions.AuthenticationException`. We should consider moving this to `core`.

Also depends on other model classes especially `ResultSet`, again candidates for moving to core.

### `auth-table-based-service` module

Alternate implementation of the AuthenticationService and AuthorizationService defined in authnz. This would continue to be part of the coordinator node. No changes required for v2. Similar comments as on `auth-jwt-service` module apply here.

### `stargate-starter` module

Nothing else depends on this, but treating as common because it is currently used to start everything. 

### `service-common` module (new)

TBD contents

### `rest-api` module

Re-implement `io.stargate.web.restapi.dao.RestDB` (previously known as `AuthenticatedDb`) to call gRPC stub instead of `DataStore` from `persistence-api`. Then address any remaining dependencies on `persistence-api`.

### `document-api` module (new)

Complete refactoring to separate the document API out of v1. 

* There are a couple of model classes that might need to go to service-common.

Re-implement `io.stargate.web.docsapi.DocumentDB` to to call gRPC stub instead of DataStore from `persistence-api`. Then address any remaining dependencies on `persistence-api`.

### `grpc` module

At first this can remain in the coordinator. This can be copied as the initial implementation of the bridge. Then re-implement `io.stargate.grpc.service.*` to call bridge gRPC stub instead of `Persistence` from `persistence-api`. We can then address any remaining dependencies on `persistence-api`.

### `testing` module

Integration tests. Will require some changes to deployment in order to run Service APIs. Will be a good opportunity to run tests that leverage only the API under test.

### `testing-services` module

This provides a self-contained jar that facilitates testing.

**Comment by shashankbrgowda** [comment]
Very helpful. Please keep us posted on the v2 design decisions and new repo details. thanks

**Reply by jeffreyscarpenter** [reply]
Yes, we will continue to update. We're wrapping up work on our first milestone for v2 - REST API factored out as its own microservice. We hope to cut our first v2 snapshot release soon and publish a blog with what we've learned so far.

**Reply by shashankbrgowda** [reply]
May I know the reason why Dropwizard is being used for new REST API microservice instead of modern frameworks like Quarkus or Micronaut ?

**Reply by jeffreyscarpenter** [reply]
We do intend to migrate to a modern framework, but haven't started that work yet. I was chatting with @tatu-at-datastax and @ivansenic yesterday about this very topic, so your timing for asking is very good :). 

I've started a conversation on the discussions where we'd love to have your input, we're starting with requirements: https://github.com/stargate/stargate/discussions/1526
